src/ci/pgo-windows.sh

```diff
@@ -0,0 +1,156 @@
+
+# rustc-perf version from 2022-05-18
+PERF_COMMIT=f66cc8f3e04392b0e2fd811f21fd1ece6ebaded3
+retry curl -LS -o $PGO_TMP/perf.zip \
+    https://github.com/rust-lang/rustc-perf/archive/$PERF_COMMIT.zip && \
+    cd $PGO_TMP && unzip -q perf.zip && \
+    mv rustc-perf-$PERF_COMMIT $RUSTC_PERF && \
+    rm perf.zip
+
+cd $RUSTC_PERF
+
+# Build rustc-perf's collector ahead of time, which is needed to make sure the rustc-fake
+# binary used by the collector is present.
+RUSTC=$RUSTC_STAGE_0 \
+RUSTC_BOOTSTRAP=1 \
+$CARGO_STAGE_0 build -p collector
+
+# Here we're profiling LLVM, so we only care about `Debug` and `Opt`, because we want to stress
+# codegen. We also profile some of the most prolific crates.
+gather_profiles "Debug,Opt" "Full" \
+    "syn-1.0.89,cargo-0.60.0,serde-1.0.136,ripgrep-13.0.0,regex-1.5.5,clap-3.1.6,hyper-0.14.18"
+
+LLVM_PROFILE_MERGED_FILE=$PGO_TMP/llvm-pgo.profdata
+
+# Merge the profile data we gathered for LLVM
+# Note that this uses the profdata from the clang we used to build LLVM,
+# which likely has a different version than our in-tree clang.
+$DOWNLOADED_LLVM/bin/llvm-profdata merge -o ${LLVM_PROFILE_MERGED_FILE} ${LLVM_PROFILE_DIRECTORY_ROOT}
+
+echo "LLVM PGO statistics"
+du -sh ${LLVM_PROFILE_MERGED_FILE}
+du -sh ${LLVM_PROFILE_DIRECTORY_ROOT}
+echo "Profile file count"
+find ${LLVM_PROFILE_DIRECTORY_ROOT} -type f | wc -l
+
+rm -r $LLVM_PROFILE_DIRECTORY_ROOT
+
+# Rustbuild currently doesn't support rebuilding LLVM when PGO options
+# change (or any other llvm-related options); so just clear out the relevant
+# directories ourselves.
+rm -r $BUILD/llvm $BUILD/lld
+
+# Okay, LLVM profiling is done, switch to rustc PGO.
+
+# The path has to be absolute
+RUSTC_PROFILE_DIRECTORY_ROOT=$PGO_TMP/rustc-pgo
+
+python ./x.py build --target=$PGO_HOST --host=$PGO_HOST \
+    --stage 2 library/std \
+    --rust-profile-generate=${RUSTC_PROFILE_DIRECTORY_ROOT}
+
+# Here we're profiling the `rustc` frontend, so we also include `Check`.
+# The benchmark set includes various stress tests that put the frontend under pressure.
+gather_profiles \
+    "Check,Debug,Opt" "All" \
+    "externs,ctfe-stress-5,cargo-0.60.0,token-stream-stress,match-stress,tuple-stress,diesel-1.4.8,bitmaps-3.1.0"
+
+RUSTC_PROFILE_MERGED_FILE=$PGO_TMP/rustc-pgo.profdata
+
+# Merge the profile data we gathered
+$BUILD/llvm/bin/llvm-profdata \
+    merge -o ${RUSTC_PROFILE_MERGED_FILE} ${RUSTC_PROFILE_DIRECTORY_ROOT}
+
+echo "Rustc PGO statistics"
+du -sh ${RUSTC_PROFILE_MERGED_FILE}
+
+rm -r $RUSTC_PROFILE_DIRECTORY_ROOT
+
+# Rustbuild currently doesn't support rebuilding LLVM when PGO options
+# change (or any other llvm-related options); so just clear out the relevant
+# directories ourselves.
+rm -r $BUILD/llvm $BUILD/lld
+
+# This produces the actual final set of artifacts, using both the LLVM and rustc
+# collected profiling data.
+$@ \
+    --rust-profile-use=${RUSTC_PROFILE_MERGED_FILE} \
+    --llvm-profile-use=${LLVM_PROFILE_MERGED_FILE}
```

Comment:
I'm not a huge fan of splitting this script into two separate files -- it looks like there's quite a bit in common between the two, and we should be able to share much of the code, with a few targeted ifs. I'm totally OK with us having an argument to the script that's either windows or unix so we don't need to detect the platform beyond that.

It looks like most of the differences are due to the Windows script being *more* generic (e.g., using env variables instead of hardcoding paths), so hopefully it's not too hard to unify the main body.

Reply:
We could have it as an argument, but I think we could also use the script helper functions like `isWindows`.

I was trying to make the two scripts more alike for an eventual merger but I can do that now, sure.

Reply:
I've pushed a few commits to do that. I'll do a couple of CI runs to ensure I haven't broken anything, and we can then iterate on these arg vs helper details if need be.

Reply:
Helper should be totally fine.